### PR TITLE
add services to the container using AddMongoIdentity extension method

### PR DESCRIPTION
**Pull request by rethabile-dev**

### CODE REVIEW

**Comment by tugberkugurlu** [comment]
Thanks for this! 🎉

File: src/AspNetCore.Identity.MongoDB/project.json [hunk 1]
```diff
@@ -10,9 +10,9 @@
         "tags": ["identity", "mongodb", "mongo", "aspnetcore", "auth"]
     },
     "dependencies": {
-        "Microsoft.AspNetCore.Identity": "1.0.0",
-        "Microsoft.Extensions.Logging.Abstractions": "1.0.0",
-        "MongoDB.Driver": "2.3.0"
+        "Microsoft.AspNetCore.Identity": "1.1.0",
+        "Microsoft.Extensions.Logging.Abstractions":"1.1.0",
+        "MongoDB.Driver": "2.4.1"
```
Comment (tugberkugurlu) [inline src/AspNetCore.Identity.MongoDB/project.json]: Let's not change the versions in this PR if we don't have to. I am happy to get these version changes in in a separate PR. Would you be OK to send these version changes in a separate PR?

Comment (tugberkugurlu) [reply src/AspNetCore.Identity.MongoDB/project.json]: Also, we should rely on `Microsoft.AspNetCore.Identity` v1.0.0 I think as I would like to support that, too.

Comment (tugberkugurlu) [reply src/AspNetCore.Identity.MongoDB/project.json]: same goes for `Microsoft.Extensions.Logging.Abstractions`.

File: src/AspNetCore.Identity.MongoDB/UserClaimsPrincipalFactory.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
+﻿using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Identity;
+using Microsoft.Extensions.Options;
+using System;
+using System.Security.Claims;
+using System.Threading.Tasks;
+
+namespace AspNetCore.Identity.MongoDB
+{
+    public class UserClaimsPrincipalFactory<TUser> : IUserClaimsPrincipalFactory<TUser>
```
Comment (tugberkugurlu) [inline src/AspNetCore.Identity.MongoDB/UserClaimsPrincipalFactory.cs]: Can we rename this to something like `ClaimsBasedUserClaimsPrincipalFactory` so that it wouldn't clash with [the one which is in Identity library itself](https://github.com/aspnet/Identity/blob/1.0.0/src/Microsoft.AspNetCore.Identity/UserClaimsPrincipalFactory.cs). I know that the generic parameters makes it distinguishable but it could still cause some confusion.

Comment (tugberkugurlu) [reply src/AspNetCore.Identity.MongoDB/UserClaimsPrincipalFactory.cs]: As a side note, this is related to https://github.com/aspnet/Identity/pull/952 (just to keep reference to that issue, no action required on this PR for that).

File: src/AspNetCore.Identity.MongoDB/MongoUserStore.cs [hunk 3]
```diff
@@ -12,7 +12,7 @@
 
 namespace AspNetCore.Identity.MongoDB
```
Comment (tugberkugurlu) [inline src/AspNetCore.Identity.MongoDB/MongoUserStore.cs]: all the formatting changes to this file is unnecessary in context of this PR. Can you revert these changes if possible?

File: samples/IdentitySample.Mvc/Properties/launchSettings.json [hunk 4]
```diff
@@ -16,7 +16,9 @@
       }
     },
     "web": {
-      "commandName": "web",
+      "commandName": "Project",
```
Comment (tugberkugurlu) [inline samples/IdentitySample.Mvc/Properties/launchSettings.json:19]: what is the difference between `web` and `Project` here?

File: src/AspNetCore.Identity.MongoDB/project.json (hunk 1 above)
Comment (rethabile-dev) [reply src/AspNetCore.Identity.MongoDB/project.json]: I'm happy to change them back.

File: src/AspNetCore.Identity.MongoDB/UserClaimsPrincipalFactory.cs (hunk 2 above)
Comment (rethabile-dev) [reply src/AspNetCore.Identity.MongoDB/UserClaimsPrincipalFactory.cs]: I see. Even more confusion if `TRole` is added at a later stage.

File: src/AspNetCore.Identity.MongoDB/MongoUserStore.cs (hunk 3 above)
Comment (rethabile-dev) [reply src/AspNetCore.Identity.MongoDB/MongoUserStore.cs]: cool. on it!

File: samples/IdentitySample.Mvc/Properties/launchSettings.json (hunk 4 above)
Comment (rethabile-dev) [reply samples/IdentitySample.Mvc/Properties/launchSettings.json:19]: `Project` is  for running application in the console host. To see the log messages on the console.

`web` is the name you click to switch between `IIS Express` and `console host`.

web

File: src/AspNetCore.Identity.MongoDB/project.json (hunk 1 above)
Comment (tugberkugurlu) [reply src/AspNetCore.Identity.MongoDB/project.json]: @rm2k thanks a lot!

File: src/AspNetCore.Identity.MongoDB/MongoUserStore.cs [hunk 5]
```diff
@@ -71,7 +71,7 @@ public MongoUserStore(IMongoDatabase database, ILoggerFactory loggerFactory, str
 
             cancellationToken.ThrowIfCancellationRequested();
 
-            await _usersCollection.InsertOneAsync(user, cancellationToken).ConfigureAwait(false);
+            await _usersCollection.InsertOneAsync(user, cancellationToken: cancellationToken).ConfigureAwait(false);
```
Comment (rethabile-dev) [inline src/AspNetCore.Identity.MongoDB/MongoUserStore.cs]: For this one I'm not sure if you are aware of this change? This is because the method is now [obsolete](https://github.com/mongodb/mongo-csharp-driver/blob/master/src/MongoDB.Driver/IMongoCollection.cs#L338).

File: samples/IdentitySample.Mvc/Properties/launchSettings.json (hunk 4 above)
Comment (rethabile-dev) [reply samples/IdentitySample.Mvc/Properties/launchSettings.json:19]: oh.I see it's a visual studio code solution. I'm not sure if the original behavior will still hold after this change. I haven't tried vs code debugger.

File: src/AspNetCore.Identity.MongoDB/UserClaimsPrincipalFactory.cs (hunk 2 above)
Comment (rethabile-dev) [reply src/AspNetCore.Identity.MongoDB/UserClaimsPrincipalFactory.cs]: How does `MongoUserClaimsPrincipalFactory` sound?

File: src/AspNetCore.Identity.MongoDB/MongoIdentityBuilder.cs [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        public Type UserType { get; private set; }
+        public IServiceCollection Services { get; private set; }
+
+        private MongoIdentityBuilder AddScoped(Type serviceType, Type concreteType)
+        {
+            Services.AddScoped(serviceType, concreteType);
+            return this;
+        }
+
+        public virtual MongoIdentityBuilder AddUserValidator<T>() where T : class
```
Comment (tugberkugurlu) [inline src/AspNetCore.Identity.MongoDB/MongoIdentityBuilder.cs:25]: This doesn't seem to be used and several other members of this class. Why do we need this (several other similar ones) as a public member?

Comment (rethabile-dev) [reply src/AspNetCore.Identity.MongoDB/MongoIdentityBuilder.cs:25]: The builder is there to allow flexibility in adding your own implementations. Let's take password validator, for instance, if you want extra password validation you just provide your own implementation and hook it along. Then, Identity will just validate against all provided validators: https://github.com/aspnet/Identity/blob/dev/src/Microsoft.AspNetCore.Identity/UserManager.cs#L2396-L2413

**Comment by tugberkugurlu** [comment]
@rm2k there are a few outstanding questions with this PR. Do you want to have a look at them? This is something I want to get it in but with minimum public APIs.

**Comment by rm2k** [comment]
@tugberkugurlu sorry which questions are you referring to?

**Comment by rethabile-dev** [comment]
@tugberkugurlu did my comments come through? what's your take on them?

**Comment by tugberkugurlu** [comment]
Sorry, it has been a while. While I was busy with other stuff, v2.0.0 has been released and it seems like v1 is not compatible with that. So, move to v2 makes more sense now.

It also seems that using the providers w/o roles is easier now. `UserClaimsPrincipalFactory<TUser>` is provided out of the box but `AddIdentity<TUser>` is not. Would you like to add something smilar to https://github.com/tugberkugurlu/AspNetCore.Identity.MongoDB/blob/a88391b52fb9dc6c615461b9d53f797461b85c55/samples/IdentitySample.Mvc/MongoIdentityServiceCollectionExtensions.cs and PR that instead?

Also, some extension methods to add the store to the DI system would be nice if you want to PR that as well.

**Comment by tugberkugurlu** [comment]
@rm2k I will close this one but let me know about https://github.com/tugberkugurlu/AspNetCore.Identity.MongoDB/pull/34#issuecomment-328307809

**Comment by rethabile-dev** [comment]
@tugberkugurlu  Sure. I'll look into your recommendations.
